solidity/contracts/bridge/Bridge.sol

```diff
@@ -773,4 +796,191 @@ contract Bridge is Ownable {
+                        request.redeemer,
+                        request.redeemerOutputHash,
+                        redeemer.walletPubKeyHash,
+                        redeemer.amount
+                    )
+                )
+            );
+
+        require(
+            redemptionRequests[redemptionRequestKey] == uint32(0),
+            "Redemption request already registered"
+        );
+        /* solhint-disable-next-line not-rely-on-time */
+        redemptionRequests[redemptionRequestKey] = uint32(block.timestamp);
+
+        emit RedemptionRequested(
+            request.redeemer,
+            request.redeemerOutputHash,
+            redeemer.walletPubKeyHash,
+            redeemer.amount
+        );
+
+        bank.transferBalanceFrom(
+            request.redeemer,
+            address(this),
+            request.amount
+        );
+    }
+
+    // TODO: Documentation.
+    function submitRedemptionProof(
+        BitcoinTx.Info calldata redemptionTx,
+        BitcoinTx.Proof calldata redemptionProof,
+        BitcoinTx.UTXO calldata mainUtxo,
+        bytes20 walletPubKeyHash
+    ) external {
+        // The actual transaction proof is performed here. After that point, we
+        // can assume the transaction happened on Bitcoin chain and has
+        // a sufficient number of confirmations as determined by
+        // `txProofDifficultyFactor` constant.
+        bytes32 redemptionTxHash =
+            validateTxProof(redemptionTx, redemptionProof);
+
+        // Assert that main UTXO for passed wallet exists in storage.
+        bytes32 mainUtxoHash = mainUtxos[walletPubKeyHash];
+        require(mainUtxoHash != bytes32(0), "No main UTXO for given wallet");
+
+        // Assert that passed main UTXO parameter is the same as in storage and
+        // can be used for further processing.
+        require(
+            keccak256(
+                abi.encodePacked(
+                    mainUtxo.txHash,
+                    mainUtxo.txOutputIndex,
+                    mainUtxo.txOutputValue
+                )
+            ) == mainUtxoHash,
+            "Invalid main UTXO data"
+        );
+
+        // Assert that the single redemption transaction input actually
+        // refers to the wallet's main UTXO.
+        (bytes32 redemptionTxInputTxHash, uint32 redemptionTxInputTxIndex) =
+            processRedemptionTxInput(redemptionTx.inputVector);
+        require(
+            mainUtxo.txHash == redemptionTxInputTxHash &&
+                mainUtxo.txOutputIndex == redemptionTxInputTxIndex,
+            "Redemption transaction input must point to the wallet's main UTXO"
+        );
+
+        // Process redemption transaction outputs and extract the change output
+        // and all information needed to perform deposit bookkeeping.
+        (uint32 redemptionTxChangeIndex, uint64 redemptionTxChangeValue) =
+            processRedemptionTxOutputs(redemptionTx.outputVector);
+
+        // TODO: Fee processing.
```

Comment:
Say we have 3 redemption requests and the fee is 5 satoshi. We will have 4 outputs in the transaction: 3 outputs for redemptions and 1 output for the main UTXO. 

I would divide the fee evenly with the last redemption request paying more in case there is a change ([example here](https://github.com/threshold-network/solidity-contracts/blob/main/contracts/vending/VendingMachine.sol#L162-L164)) and having the main UTXO paying no fee:
- first redemption request: fee 1 sat
- second redemption request: fee 1 sat
- third redemption request: fee 3 sat
- main UTXO: fee 0 sat

Regarding max fee, the RFC 1 says:

> Note: We do not allow users to specify a max btc fee. When users deposit, they’re agreeing to be swept at whatever fee the operators decide is appropriate (based on https://blockstream.info/api/fee-estimates). Operators cannot pick a starting fee higher than base_sweeping_fee_max

This makes sense but we could have a global, governable `max_fee` for both deposit and redemption that would be set to some pretty high value. This would be on-chain protection against a malicious wallet being a Bitcoin miner and taking 99% fee from the deposited/redemeed value.

If the max fee was exceeded, it will not be possible to submit sweep/redemption proof and another function for fee fraud would be exposed. Not sure though what that function would do yet. Maybe request the DAO to slash/resolve the situation? cc @beaushinkle

Reply:
I think it's no longer relevant for redemption since we just validate whether the given output value claiming to fulfill a given redemption request falls within the request's acceptable range (between `requestedAmount - treasuryFee` and `requestedAmount - treasuryFee - txMaxFee`). On the other hand, we'll employ that mechanism for sweep fee accounting.